ethereum/rpc/types/block.go

```diff
@@ -103,3 +106,73 @@ func (bn BlockNumber) TmHeight() *int64 {
 	height := bn.Int64()
 	return &height
 }
+
+// BlockNumberOrHash represents a block number or a block hash.
+type BlockNumberOrHash struct {
+	BlockNumber *BlockNumber `json:"blockNumber,omitempty"`
+	BlockHash   *common.Hash `json:"blockHash,omitempty"`
+}
+
+func (bnh *BlockNumberOrHash) UnmarshalJSON(data []byte) error {
+	type erased BlockNumberOrHash
+	e := erased{}
+	err := json.Unmarshal(data, &e)
+	if err == nil {
+		return bnh.checkUnmarshal(BlockNumberOrHash(e))
+	}
+	var input string
+	err = json.Unmarshal(data, &input)
+	if err != nil {
+		return err
+	}
+	err = bnh.decodeFromString(input)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bnh *BlockNumberOrHash) checkUnmarshal(e BlockNumberOrHash) error {
+	if e.BlockNumber != nil && e.BlockHash != nil {
+		return fmt.Errorf("cannot specify both BlockHash and BlockNumber, choose one or the other")
+	}
+	bnh.BlockNumber = e.BlockNumber
+	bnh.BlockHash = e.BlockHash
+	return nil
+}
+
+func (bnh *BlockNumberOrHash) decodeFromString(input string) error {
+	switch input {
+	case "earliest":
+		bn := EthEarliestBlockNumber
+		bnh.BlockNumber = &bn
+	case "latest":
+		bn := EthLatestBlockNumber
+		bnh.BlockNumber = &bn
+	case "pending":
+		bn := EthPendingBlockNumber
+		bnh.BlockNumber = &bn
+	default:
```

Comment:
can you add code comments on the default case?

Reply:
ok